keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -161,273 +122,289 @@ consistent with the syntax supported by any of the common proxy vendors:
-| `[xyz]`    | character class         |
-| `[^xyz]`   | negated character class |
-| `x*`       | 0 or more x's           |
-| `x+`       | 1 or more x's           |
-| `xy`       | x followed by y         |
-| `x|y`      | x or y (prefer x)       |
-| `(abc)`    | grouping                |
+  // Defaults to "/" if empty.
+  //
+  // +Optional
+  Path string
 
-Maintaining a regular expression subset is likely not worth the complexity and
-is likely impossible across the [many implementations][regex-survey].
+  // PathType determines the interpretation of the Path
+  // matching. PathType can be one of the following values:
+  //
+  // Exact  - matches the URL path exactly.
+  // Prefix - matches based on a URL path prefix split
+  //   by '/'. [insert description of semantics described below]
+  //
+  // ImplementationSpecific - interpretation of the Path
+  //   matching is up to the IngressClass. Not all
+  //
+  // +Optional
+  PathType string
+  ...
+ }
+ ```
+
+Interoperability between versions will be as follows:
+
+1. v1beta1 will default to `ImplementationSpecific`.
+1. v1 will default to `Prefix`.
+
+Note: default value are permitted between API versions
+([reference][api-conv-versions]).
+
+[api-conv-versions]: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#defaulting
+
+##### Path matching semantics
+
+For non-`ImplementationSpecific` paths:
+
+1. Let `[p_1, p_2, ..., p_n]` be the list of Paths for a specific host.
+1. Every Path `p_i` must be syntactically valid:
+    1. Must begin with the `'/'` character (relative paths are not allowed by [RFC-7230][rfc7230]).
+    1. Must not contain consecutive `'/'` characters (e.g. `/foo///`, `//`).
+1. A trailing `'/'` character in the Path is ignored, e.g. `/abc` and `/abc/`
+   specify the same match. The rest of the discussion below assumes
+   trailing `'/'` are removed from the paths.
+1. If there is more than one potential match:
+   1. `Exact` match is preferred to a `Prefix` match.
+   1. For multiple prefix matches, the longest Path `p_i` will be the
+      matching path.
+   1. If an `ImplementationSpecific` match exists in the spec, then the
+      preference depends on the implementation.
+1. If there is no matching path, then the `defaultBackend` for the host will be
+   used.
+1. If there is not match for the host, then the overall `defaultBackend` for
+   the Ingress will be selected.
+
+##### `Exact` match
+
+Path must be exactly the same as the request path.
+
+##### `Prefix` match
+
+Matching is done on a path element by element basis. A path element refers
+is the list of labels in the path split by the `'/'` separator. A request
+is a match for path `p` if every `p` is an element-wise prefix of `p` of the
+request path. Note that if the last element of the path is a substring of
+the last element in request path, it
+is *not* a match (e.g. `/foo/bar` matches `/foo/bar/baz`, but does not
```

Comment:
Again, is this the common understanding among proxies?

Reply:
This looks to be implementable (without too much pain) for nginx, envoy, haproxy and AWS, GCE.

Reply:
We may want to rename "Prefix" though to avoid confusion with StringPrefix.